### PR TITLE
[RFC- Don't merge] Alternate byte implementation for non VC / pre c++17 compiler

### PR DESCRIPTION
**Pull request by MikeGitb**
As discussed in issue #313, the current implementation of byte leads to undefined behavior when data is accessed through a pointer to `gsl::byte` (contrary to what we are used with e.g. `unsigned char*`). Due to strict alias assumption in g++ and clang++ 's optimizer, this can demonstrably lead to errors in certain cases, while VC++ seems not to be affected.

This PR implements two alternative versions of `gsl::byte`
- As a simple typedef for `unsigned char`, which gets rid of the UB, but also looses the type safety properties of the enum version.
- as a struct wrapping a single unsigned char as used in the gsl-lite by @martinmoene. 

Warning: I believe the second version is still undefined behavior and it leads to failures on my system. I can't reproduce those failures with the gsl-lite however which - I believe - has the same implementation. So maybe those failure are only caused by a bug in my implementation.

### CODE REVIEW

File: gsl/gsl_byte [hunk 1]
```diff
@@ -36,8 +36,167 @@
...
+    explicit operator unsigned char&() { return v; }
+    explicit operator const unsigned char&() const { return v; }
+    friend bool operator==(byte l, byte r) { return l.v == r.v; }
+    friend bool operator!=(byte l, byte r) { return l.v != r.v; }
+};
+
+template <class IntegerType, class = std::enable_if_t<std::is_integral<IntegerType>::value>>
+inline constexpr byte& operator<<=(byte& b, IntegerType shift) noexcept
+{
+    return b.v <<= shift, b;
```
Comment (martinmoene) [inline gsl/gsl_byte]: Funny, initially was going to write it as here, but in the end chose `b.v <<= shift; return b;`

**Comment by martinmoene** [comment]
When compiling `byte_tests` via CMake with g++ (GCC) 5.2.0 on Windows it selects `GSL_BYTE_TYPE_ENUM` and fails the `aliasing` test.

In the same situation, but with `GSL_USE_BYTE_TYPE` defined as `GSL_BYTE_TYPE_STRUCT` the aliasing test passes.

File: gsl/gsl_byte (hunk 1 above)
Comment (MikeGitb) [inline gsl/gsl_byte]: I like your version better, but tried to make it compatible with c++11 constexpr rules (which is somewhat pointless, as the ms-gls doesn't work with c++11 anyway). Do you have any Idea, why the unit test fails here but not in your gsl-lite?

**Comment by MikeGitb** [comment]
@martinmoene: Sorry, my mistake - I tested the different values for ´GSL_USE_BYTE_TYPE` prior to creating the PR and then accidentally committed the default enum version instead of the struct version.
Strangely enough, with gcc 6.1 on fedora it still fails for me.

File: gsl/gsl_byte (hunk 1 above)
Comment (martinmoene) [inline gsl/gsl_byte]: Did you specify `@` or `"[.issue #34]"` (with or  without `--pass`) as below to actually invoke the alias test for gsl-lite?

>prompt> gsl-lite-cpp14.t.exe --pass "[.issue #34]"
path/to/issue.t.cpp:47: passed: byte: aliasing rules lead to undefined behaviour when using enum class[.issue #34](GSL #313): 14 == F::f( i, reinterpret_cast<gsl::byte&>( i ) ) for 14 == 14
All 1 selected test passed.

(Did note the s/test/tests/)

Comment (MikeGitb) [inline gsl/gsl_byte]: no, I just called `make test`  and assumed that would call all tests. But yes, even after explicitly selecting the test as shown by you, it passes.

**Comment by MikeGitb** [comment]
Rebased to master and integrated fix to select struct based byte by default on non-VC++ compiler

**Comment by MikeGitb** [comment]
Ok, the new travis tests also show that the struct version doesn't reliably work on gcc 6 but on all other. So I guess the question is if 

1.  I should keep all three versions and use the "best" one (`enum` > `struct` > `unsigned int`) that works on a certain compiler.
2. The same as 1. but only with the enum and unsigned int versions
3. Leave the `gsl/gsl_byte` header as it is and just accept that `gsl::byte` doesn't completely work as intended on all compilers

**Comment by gdr-at-ms** [comment]
I continue to suggest `-fno-strict-aliasing` -- it sounds like the compiler is being too aggressive even with time-honored structures.

**Comment by MikeGitb** [comment]
@gdr-at-ms: IIRC, neilmacintosh wasn't to happy with that solution. Personally, I don't care too much anymore, but if you two agree that this is the best way to address this bug, then that requirement should imho be mentioned somewhere (e.g. the readme).

**Comment by MikeGitb** [comment]
As there isn't any new input and you seem to be happy with the solution in #390 for now, I think this PR can be closed. I won't delete the branch in the near future, so if this comes up again it can be used as a basis for further exploration.

**Pull request closed without merging**